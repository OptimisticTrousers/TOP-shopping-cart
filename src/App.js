import React from "react";
import { Outlet } from "react-router-dom";
import Navbar from "./components/Navbar";
import Footer from "./components/Footer";
import { CartProvider } from "./context/Store";

export default function App() {

  return (
    <div className="container">
      <CartProvider>
        <Navbar />
        <main>
          <Outlet />
          <Footer />
        </main>
      </CartProvider>
    </div>
  );
}