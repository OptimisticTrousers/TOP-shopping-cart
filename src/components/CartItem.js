import trash from '../assets/trash.svg'
import { useAddToCart, useReduceQuantity, useRemoveFromCart} from '../context/Store'

export default function CartItem(props) {

    const {image, title, price, quantity} = props

    const addToCart = useAddToCart()

    const reduceQuantity = useReduceQuantity()

    const removeFromCart = useRemoveFromCart()

    return (
        <div className="cart-item-container" data-testid="cart-item">
            <div className="item-image">
                <img src={image} alt={title}/>
            </div>
            <div className="cart-item-details">
                <div className="item-detail">
                    <h2 className='item-title product-title'>{title}</h2>
                    <p className='item-price product-price'>${price}</p>
                </div>
                <div className="item-stock">
                    <p className="item-stock-price">${price}</p>
                    <p className="item-stock-status">In Stock</p>
                </div>
                <div className="user-input">
                    <div className="quantity">
                        <button type="button" onClick={() => addToCart(props)}>+</button>
                        <p className="quantity-display" data-testid="product-quantity">{quantity}</p>
                        <button type="button" onClick={() => reduceQuantity(props)}>-</button>
                    </div>
                    <div className="delete">
                        <img className="trash-image" src={trash} alt="trash can"/>
                        <p className="text" onClick={() => removeFromCart(props)}>Delete</p>
                    </div>
                </div>
            </div>
        </div>
    )
}