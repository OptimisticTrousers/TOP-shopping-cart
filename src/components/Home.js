import Hero from "./Hero"
import About from "./About"

export default function Home() {
    return (
        <div className="home">
            <Hero /> 
            <About />
        </div>
    )
}