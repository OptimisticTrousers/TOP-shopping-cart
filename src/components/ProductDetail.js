import { Link, useParams } from "react-router-dom"
import { useEffect, useState } from "react"
import leftArrow from '../assets/left-arrow.svg'
import { useAddToCart } from "../context/Store"

export default function ProductDetail(){

    const addToCart = useAddToCart()

    const params = useParams()

    const [details, setDetails] = useState({})

    useEffect(() => {
        fetch(`https://fakestoreapi.com/products/${params.id}`)
            .then(res => res.json())
            .then(data => setDetails(data))
    }, [])

    return (
        <div className="detail-container">
            <div className="image-detail-view">
                <img className="image-detail" src={details.image} alt={details.title}/>
            </div>
            <div className="product-details">
                <Link to="/shop">
                    <button type="button"><img className="left-arrow" src={leftArrow} alt="left arrow" /> Back To Products</button>
                </Link>
                <h1 className="product-title">{details.title}</h1>
                <p className="product-price">${details.price}</p>
                <p className="category-detail">Category: {details.category}</p>
                <p className="description-detail">{details.description}</p>
                <button onClick={() => addToCart({...details, quantity: 0})} className="add-to-cart-button" type="button">Add to Cart</button>
            </div>
        </div>
    )
}