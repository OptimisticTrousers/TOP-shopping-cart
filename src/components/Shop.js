export default function Shop() {
    return (
        <main>
            <h1>Get Clothing!</h1>
            <p>Times are tough. Liven up with our top-notch clothing!</p>
            <div className="products">

            </div>
        </main>
    )
}