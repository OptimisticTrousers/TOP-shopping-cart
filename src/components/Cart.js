import CartItem from "./CartItem"
import logo from '../assets/optimistictrousers.jpg'

export default function Cart() {

    return (
        <div className="cart-container">
            <h1 className="cart-title">Your Cart</h1>
            <div className="cart">
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>

            </div>
            <p>SUBTOTAL</p>
            <div className="buttons">
                <button type="button">Check Out</button>
                <button type="button">Back To Products</button>
            </div>
        </div>
    )
}