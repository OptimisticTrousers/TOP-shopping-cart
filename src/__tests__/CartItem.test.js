import React from 'react'
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import CartItem from '../components/CartItem';
import {BrowserRouter, Routes, Route} from 'react-router-dom'
import App from '../App'
import Home from '../components/Home'
import Shop from '../components/Shop'
import Cart from '../components/Cart'
import ProductDetail from '../components/ProductDetail'

describe('CartItem', () => {

    test('screenshot', () => {
        const {asFragment} = render(<CartItem />)

        expect(asFragment()).toMatchSnapshot()
    })

    test('user can increment quantity', async () => {

        render(
            <React.StrictMode>
                <BrowserRouter>
                    <Routes>
                        <Route path="/" element={<App />} >
                            <Route index element={<Home />} />
                            <Route path="shop" element={<Shop />} />
                            <Route path="shop/:id" element={<ProductDetail />} />
                            <Route path="cart" element={<Cart />} />
                        </Route>
                    </Routes>
                </BrowserRouter>
            </React.StrictMode>
        )

        const user = userEvent.setup()

        // Adding first product to cart
        await user.click(screen.queryByText("SHOP"))
        const firstProduct = await screen.findByAltText('Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops')
        await user.click(firstProduct)
        await user.click(screen.getByRole('button', {name: /Add to Cart/i}))

        // Adding second product to cart
        await user.click(screen.queryByText("SHOP"))
        const secondProduct = await screen.findByAltText('Mens Casual Premium Slim Fit T-Shirts')
        await user.click(secondProduct)
        await user.click(screen.getByRole('button', {name: /Add to Cart/i}))

        // Going to checkout
        const cartSVG = screen.getByTestId('cart-svg')
        await user.click(cartSVG)

        //Clicking quantity incrementor for first product only
        const [firstItemTitle] = await screen.findAllByRole('button', {name: "+"})
        await user.click(firstItemTitle)
        const firstItemQuantity = await screen.findAllByTestId('product-quantity')


        expect(firstItemQuantity).toBe("2")
    })
})