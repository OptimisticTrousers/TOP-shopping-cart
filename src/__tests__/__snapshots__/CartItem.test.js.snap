// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem screenshot 1`] = `
<DocumentFragment>
  <div
    class="cart-item-container"
    data-testid="cart-item"
  >
    <div
      class="item-image"
    >
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      />
    </div>
    <div
      class="cart-item-details"
    >
      <div
        class="item-detail"
      >
        <h2
          class="item-title product-title"
        >
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h2>
        <p
          class="item-price product-price"
        >
          $109.95
        </p>
      </div>
      <div
        class="item-stock"
      >
        <p
          class="item-stock-price"
        >
          $109.95
        </p>
        <p
          class="item-stock-status"
        >
          In Stock
        </p>
      </div>
      <div
        class="user-input"
      >
        <div
          class="quantity"
        >
          <button
            type="button"
          >
            +
          </button>
          <p
            class="quantity-display"
            data-testid="product-quantity"
          >
            1
          </p>
          <button
            type="button"
          >
            -
          </button>
        </div>
        <div
          class="delete"
        >
          <img
            alt="trash can"
            class="trash-image"
            src="trash.svg"
          />
          <p
            class="text"
          >
            Delete
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
